calculator.rb

```diff
@@ -0,0 +1,174 @@
+	modulus: ["modulus", "modulo", "mod", "remainder", "%"]
+}
+
+# Check a number inputted by the user as a string and see if it has a decimal
+# If it has a decimal, convert it to a float, otherwise, convert it to an integer
+def check_num_format(input)
+	if input.include?(".")
+		data = input.delete(",").to_f
+	else
+ 		data = input.delete(",").to_i
+	end
+
+	return data
+end
+
+# Performe a calculation using an operation and two numbers
+# Return the output and the string form of the operator
+def calculate(operation, first_num, second_num)	
+
+	first_number_data = check_num_format(first_num)
+	second_number_data = check_num_format(second_num)
+	
+	if $supported_operations[:addition].include?(operation)
+		output = first_number_data + second_number_data
+		operator = $supported_operations[:addition][-1]
+
+	elsif $supported_operations[:subtraction].include?(operation)
+	 	output = first_number_data - second_number_data
+	 	operator = $supported_operations[:subtraction][-1]
+
+	elsif $supported_operations[:multiplication].include?(operation)
+	 	output = first_number_data * second_number_data
+	 	operator = $supported_operations[:multiplication][-1]
+
+	elsif $supported_operations[:division].include?(operation)
+	 	output = first_number_data.to_f / second_number_data.to_f
+	 	operator = $supported_operations[:division][-1]
+
+	elsif $supported_operations[:exponent].include?(operation)
+	 	output = first_number_data ** second_number_data
+	 	operator = $supported_operations[:exponent][-1]
+
+	elsif $supported_operations[:modulus].include?(operation)
+	 	output = first_number_data % second_number_data
+	 	operator = $supported_operations[:modulus][-1]
+	end
+
+	return output, operator
+end
+
+# Evaluate a statement contained in parentheses and return the value
+def evaluate_parens(statement_in_parens)
+	# strip the parens out of the input
+	statement = statement_in_parens[1...-1]
+	
+	# partition the statement using a regex to find the operation
+	# feed the elements of the array into the 2 numbers and operation
+	statement_first_num = statement.partition(/[\+\-\^\*\/\%]/)[0].strip
+	statement_operation = statement.partition(/[\+\-\^\*\/\%]/)[1].strip
+	statement_second_num = statement.partition(/[\+\-\^\*\/\%]/)[2].strip	
+
+	# run the calculate function and save the returned values into variables
+	value, operator = calculate(statement_operation, statement_first_num, statement_second_num)
+	
+	# return just the output value
+	return value
+end
+
+# Ask for what operation the user wants to perform and store the input in a variable.
+puts "What operation would you like to perform?"
+operation = gets.chomp.downcase
+
+# Create a boolean variable to store whether the operation is valid or not.
+valid = false
+
+# Check for operation validity
+while valid == false
+	# If operation falls into the flattened supported operations array, stop the loop and proceed.
+	if $supported_operations.flatten(2).include?(operation)
```

Comment:
Clever use of flatten to generate a list of all valid operations. :+1: 

One thing to be cognizant of when you're using loops is that all code will execute each time you loop. So in this case, `$supported_operations.flatten` happens every time. One way around that is to call it once before you enter the loop and store it in a variable for the loop to use.

```
supported_list = $supported_operations.flatten(2)

while valid == false
  if supported_list.include?(operation)
  ...
```